gtsam/discrete/TableFactor.h

```diff
@@ -0,0 +1,333 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file TableFactor.h
+ * @date May 4, 2023
+ * @author Yoonwoo Kim
+ */
+
+#pragma once
+
+#include <gtsam/discrete/DiscreteFactor.h>
+#include <gtsam/discrete/DiscreteKey.h>
+#include <gtsam/inference/Ordering.h>
+
+#include <Eigen/Sparse>
+#include <algorithm>
+#include <memory>
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace gtsam {
+
+  class SparseDiscreteConditional;
+  class HybridValues;
+
+  /**
+   * A discrete probabilistic factor optimized for sparsity.
+   *
```

Comment:
Please explain in some detail the approach and refer to member variables below.

Reply:
I added:
Uses sparse_table_ to store only the non-zero probabilities.
Computes the assigned value for the key using the ordering which the non-zero probabilities are stored in.